### ISSUE TITLE
 * Support more dataframes 

### ISSUE THREAD

**Issue by abhijithneilabraham** [abhijithneilabraham/tableQA#23]
-Currently the feature supports csv files only. However, integrating more dataframes is easy. Go through the get_dataframe() method in data_utils.py and include support to detect the incoming file and parse the dataframe from it.

**Comment by rhamnett** [comment]
It'd be good to just be able to do something like:
```
agent = Agent(df)
```

**Comment by abhijithneilabraham** [comment]
> It'd be good to just be able to do something like:
> 
> ```
> agent = Agent(df)
> ```

Sure. We will update it asap.

**Comment by rhamnett** [comment]
@abhijithneilabraham keep up the great work, looking forward to contributing

**Comment by abhijithneilabraham** [comment]
Thanks! @rhamnett   The repo is updated with the changes you suggested, kindly go through the readme and examples and report any bugs.

**Comment by rhamnett** [comment]
Hello @abhijithneilabraham! That's great! I gave it a quick test on a sample dataset i created from my own columns and the basic functionality seems to work!

Some support for date types etc - datetime64 etc would be really good and then getting the model to understand what today/yesterday/this week/last week means from the current date.


agent.query_db("how many website visits were there in the last week?")
agent.query_db("how many orders have I had today?")
agent.query_db("how many orders have I had for the week beginning 31st August 2020?")

This kind of thing 👍

**Comment by abhijithneilabraham** [comment]
Makes sense. In fact, I was already building something like that, but the issue is to map the natural language to datetime without approaching it in a rule based way. Let me think about this and will update asap.

**Comment by rhamnett** [comment]
> Makes sense. In fact, I was already building something like that, but the issue is to map the natural language to datetime without approaching it in a rule based way. Let me think about this and will update asap.

Yes, that's whats going to make it interesting :)

**Comment by TheurgicDuke771** [comment]
Hi,
What will be the df when use MySQL db (have multiple table)?

**Comment by abhijithneilabraham** [comment]
Haven't implemeted multi-table support yet. We need to join tables in that case, it seems. Will plan that for future. Thanks 😊 .

**Comment by TheurgicDuke771** [comment]
thanks for the quick reply. Then df should be `select * from table`, right?

**Comment by abhijithneilabraham** [comment]
That's correct. Download a table as a dataframe and pass it to this.

You could also contribute a feature if you can enable something like agent=Agent(db_url, table_name). Refer `agent.py`,`data_utils.py`

**Comment by TheurgicDuke771** [comment]
Sure, will check that. And send you a pr if get anything solid.

**Comment by abhijithneilabraham** [comment]
Thanks!

**Comment by abhijithneilabraham** [comment]
> Hello @abhijithneilabraham! That's great! I gave it a quick test on a sample dataset i created from my own columns and the basic functionality seems to work!
> 
> Some support for date types etc - datetime64 etc would be really good and then getting the model to understand what today/yesterday/this week/last week means from the current date.
> 
> agent.query_db("how many website visits were there in the last week?")
> agent.query_db("how many orders have I had today?")
> agent.query_db("how many orders have I had for the week beginning 31st August 2020?")
> 
> This kind of thing 👍

This has been done too! We have a `Date` class in `column_types.py`

**Comment by dharmesh2002** [comment]
Hi Abhijit
Have used this tool, it is an excellent tool.
Have tested some of my scenarios, they are working fine. Have one query, do joins works in this tool like left join or right join

**Comment by abhijithneilabraham** [comment]
Hi @dharmesh2002 , Can you be more specific where you have tested a scenario with SQL joins?

**Comment by dharmesh2002** [comment]
Hi @abhijithneilabraham , i haven't tested sql queries, what i need is, to connect sql and than to join with tables. So that's why i was checking with you whether that is feasible?

**Comment by dharmesh2002** [comment]
@abhijithneilabraham - is it possible to add new functions like group by or having clause?

**Comment by abhijithneilabraham** [comment]
Yes. Can you add support for them?

**Comment by dharmesh2002** [comment]
@abhijithneilabraham sorry I didn’t understood, can u pls help me where to add group by function, I tried adding but it didn’t work

**Comment by abhijithneilabraham** [comment]
Can you make a pull request with your code and tag it here?

**Comment by snehithakrishna2** [comment]
> @dharmesh2002 Can you make a pull request with your code and tag it here?
hi,
@abhijithneilabraham  https://github.com/abhijithneilabraham/tableQA/pull/61

**Comment by dharmesh2002** [comment]
@abhijithneilabraham please if you can have a look on pool request we shared above

**Comment by abhijithneilabraham** [comment]
Hi, I reviewed the code. Have you tried testing it on an actual database
and see if the results suit your needs in a natural language query?

**Comment by shubhamrangate2413** [comment]
> Sure, will check that. And send you a pr if get anything solid.
@TheurgicDuke771 , I am working on a similar use case of connecting the Agent with Postgresql db. Are you able to crack anything for making a connection with db?
